File: structural/flyweight/map_rendering/typescript/map_rendering.test.ts
import { describe, beforeEach, test, expect } from '@jest/globals';
import {
  MapRenderer,
  VehicleType,
  VehicleStatus,
  VehicleIconFactory,
} from "./map_rendering";

describe("MapRendering", () => {
  let renderer: MapRenderer;

  beforeEach(() => {
    renderer = new MapRenderer();
    VehicleIconFactory.clearCache();
  });

  test("should add a new vehicle to the map", () => {
    renderer.addVehicle("car1", VehicleType.CAR);
    expect(renderer.getVehicleCount()).toBe(1);
  });

  test("should update vehicle state", () => {
    renderer.addVehicle("car1", VehicleType.CAR);
    renderer.updateVehicle(
      "car1",
      { latitude: 37.7749, longitude: -122.4194 },
      45.0,
      VehicleStatus.MOVING,
      90.0
    );
    const renderOutput = renderer.renderMap()[0];
    expect(renderOutput).toContain("car1");
    expect(renderOutput).toContain("37.7749");
    expect(renderOutput).toContain("-122.4194");
  });

  test("should share icons between vehicles of the same type", () => {
    renderer.addVehicle("car1", VehicleType.CAR);
    renderer.addVehicle("car2", VehicleType.CAR);
    renderer.addVehicle("bus1", VehicleType.BUS);

    expect(renderer.getVehicleCount()).toBe(3);
    expect(renderer.getIconCount()).toBe(2);
  });

  test("should render all vehicles on the map", () => {
    renderer.addVehicle("car1", VehicleType.CAR);
    renderer.addVehicle("bus1", VehicleType.BUS);

    renderer.updateVehicle(
      "car1",
      { latitude: 37.7749, longitude: -122.4194 },
      45.0,
      VehicleStatus.MOVING,
      90.0
    );
    renderer.updateVehicle(
      "bus1",
      { latitude: 37.7749, longitude: -122.4194 },
      30.0,
      VehicleStatus.STOPPED,
      180.0
    );

    const renderOutputs = renderer.renderMap();
    expect(renderOutputs).toHaveLength(2);
    expect(renderOutputs.some(output => output.includes("car1"))).toBe(true);
    expect(renderOutputs.some(output => output.includes("bus1"))).toBe(true);
  });
});

describe("VehicleIconFactory", () => {
  beforeEach(() => {
    VehicleIconFactory.clearCache();
  });

  test("should create and return icon for new vehicle type", () => {
    const icon = VehicleIconFactory.getIcon(VehicleType.CAR);
    expect(icon.type).toBe(VehicleType.CAR);
    expect(icon.imageData).toBe("icon_data_for_car");
    expect(icon.width).toBe(32);
    expect(icon.height).toBe(32);
  });

  test("should return the same icon instance for the same vehicle type", () => {
    const icon1 = VehicleIconFactory.getIcon(VehicleType.CAR);
    const icon2 = VehicleIconFactory.getIcon(VehicleType.CAR);
    expect(icon1).toBe(icon2);
  });

  test("should return different icon instances for different vehicle types", () => {
    const carIcon = VehicleIconFactory.getIcon(VehicleType.CAR);
    const busIcon = VehicleIconFactory.getIcon(VehicleType.BUS);
    expect(carIcon).not.toBe(busIcon);
    expect(carIcon.type).toBe(VehicleType.CAR);
    expect(busIcon.type).toBe(VehicleType.BUS);
  });

  test("should track the number of unique icons", () => {
    VehicleIconFactory.getIcon(VehicleType.CAR);
    VehicleIconFactory.getIcon(VehicleType.CAR); // Same type
    VehicleIconFactory.getIcon(VehicleType.BUS);
    expect(VehicleIconFactory.getIconCount()).toBe(2);
  });

  test("should clear the icon cache", () => {
    VehicleIconFactory.getIcon(VehicleType.CAR);
    VehicleIconFactory.getIcon(VehicleType.BUS);
    VehicleIconFactory.clearCache();
    expect(VehicleIconFactory.getIconCount()).toBe(0);
  });
});

File: structural/flyweight/map_rendering/go/map_rendering/map_rendering_test.go
package map_rendering

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMapRendering(t *testing.T) {
	// Create a new map renderer
	renderer := NewMapRenderer()

	// Add some vehicles
	renderer.AddVehicle("car1", Car)
	renderer.AddVehicle("bus1", Bus)
	renderer.AddVehicle("truck1", Truck)
	renderer.AddVehicle("car2", Car) // Same type as car1

	// Update vehicle states
	renderer.UpdateVehicle(
		"car1",
		Position{37.7749, -122.4194},
		60,
		Moving,
		90,
	)

	renderer.UpdateVehicle(
		"bus1",
		Position{37.7833, -122.4167},
		30,
		Stopped,
		180,
	)

	renderer.UpdateVehicle(
		"truck1",
		Position{37.7750, -122.4183},
		45,
		Moving,
		270,
	)

	renderer.UpdateVehicle(
		"car2",
		Position{37.7740, -122.4200},
		55,
		Moving,
		0,
	)

	// Render the map
	renderedMap := renderer.RenderMap()
	assert.Len(t, renderedMap, 4, "Should render all vehicles")

	// Check memory efficiency
	assert.Equal(t, 4, renderer.GetVehicleCount(), "Should have 4 vehicles")
	assert.Equal(t, 3, renderer.GetIconCount(), "Should have 3 unique icons (car, bus, truck)")

	// Test icon sharing
	car1 := renderer.vehicles["car1"]
	car2 := renderer.vehicles["car2"]
	assert.Equal(t, car1.icon, car2.icon, "Cars should share the same icon instance")
}

func TestVehicleIconFactory(t *testing.T) {
	factory := NewVehicleIconFactory()

	// Get icons for different vehicle types
	carIcon1 := factory.GetIcon(Car)
	carIcon2 := factory.GetIcon(Car)
	busIcon := factory.GetIcon(Bus)

	// Test icon sharing
	assert.Equal(t, carIcon1, carIcon2, "Same vehicle type should return the same icon instance")
	assert.NotEqual(t, carIcon1, busIcon, "Different vehicle types should return different icon instances")

	// Test icon count
	assert.Equal(t, 2, factory.GetIconCount(), "Should have 2 unique icons")

	// Test cache clearing
	factory.ClearCache()
	assert.Equal(t, 0, factory.GetIconCount(), "Cache should be empty after clearing")
} 